File: 算法/代码随想录算法PDF/00调试源代码/102.二叉树的层序遍历.py
#
# @lc app=leetcode.cn id=102 lang=python3
#
# [102] 二叉树的层序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        def order(node, res, layer):
            if node == None:
                return
            if len(res) == layer:
                res.append([])
            res[layer].append(node.val)
            order(node.left, res, layer + 1)
            order(node.right, res, layer + 1)

        res = []
        layer = 0
        order(root, res, layer)
        return res
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/102.二叉树的层序遍历.cpp
/*
 * @lc app=leetcode.cn id=102 lang=cpp
 *
 * [102] 二叉树的层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void order(TreeNode* node, vector<vector<int>>& res, int layer) {
        if (node == nullptr) return;
        if (res.size() == layer) res.push_back(vector<int>());  // 开始了新的一层
        res[layer].push_back(node->val);
        order(node->left, res, layer + 1);
        order(node->right, res, layer + 1);
    }
    vector<vector<int>> levelOrder(TreeNode* root) {
        int layer = 0;  // 层数，用于确认遍历到节点值放的位置
        vector<vector<int>> res;
        order(root, res, layer);
        return res;
    }
};
// @lc code=end

